src/browser/modules/D3Visualization/lib/visualization/components/Visualization.ts

```diff
@@ -17,216 +17,113 @@
-      node.hoverFixed = false
-      node.fx = null
-      node.fy = null
-    }
-    return viz.trigger('nodeMouseOut', node)
-  }
-
-  const onRelMouseOver = (_event: Event, rel: Relationship) =>
-    viz.trigger('relMouseOver', rel)
-
-  const onRelMouseOut = (_event: Event, rel: Relationship) =>
-    viz.trigger('relMouseOut', rel)
-
-  const handleZoomOnShiftScroll = (e: WheelEvent) => {
-    if (e.shiftKey) {
-      e.preventDefault()
-
-      const delta =
-        -e.deltaY * (e.deltaMode === 1 ? 0.05 : e.deltaMode ? 1 : 0.002)
-
-      return zoomBehavior.scaleBy(baseGroup, 1 + delta)
-    }
-  }
-
-  const zoomed = function (
-    e: D3ZoomEvent<SVGGElement, unknown>
-  ): Transition<SVGGElement, unknown, null, undefined> {
-    draw = true
-
-    if (isZoomClick) {
-      // When clicking on the zoom buttons show a longer animation than when
-      // using the scroll wheel.
-      isZoomClick = false
-
-      return container
-        .transition()
-        .duration(400)
-        .ease(easeCubic)
-        .attr('transform', String(e.transform))
-    }
-
-    return container
-      .transition()
-      .duration(20)
-      .attr('transform', String(e.transform))
-  }
-
-  const zoomBehavior = d3Zoom<SVGGElement, unknown>()
-    .scaleExtent([0.1, 2])
-    .on('zoom', zoomed)
-
-  let isZoomingIn = true
-
-  viz.zoomInClick = function () {
-    isZoomingIn = true
-    return zoomClick(this)
-  }
-
-  viz.zoomOutClick = function () {
-    isZoomingIn = false
-    return zoomClick(this)
+    this.root = d3Select(element)
+
+    // Remove the base group element when re-creating the visualization
+    this.root.selectAll('g').remove()
+    this.baseGroup = this.root.append('g').attr('transform', 'translate(0,0)')
+
+    this.rect = this.baseGroup
+      .append('rect')
+      .style('fill', 'none')
+      .style('pointer-events', 'all')
+      // Make the rect cover the whole surface, center of the svg viewbox is in (0,0)
+      .attr('x', () => -Math.floor(measureSize().width / 2))
+      .attr('y', () => -Math.floor(measureSize().height / 2))
+      .attr('width', '100%')
+      .attr('height', '100%')
+      .attr('transform', 'scale(1)')
+      // Background click event
+      // Check if panning is ongoing
+      .on('click', () => {
+        if (!this.draw) {
+          return this.trigger('canvasClicked')
+        }
+      })
+
+    this.container = this.baseGroup.append('g')
+    this.geometry = new GraphGeometry(style)
+
+    this.zoomBehavior = d3Zoom<SVGGElement, unknown>()
+      .scaleExtent(ZOOM_SCALE_EXTENT)
+      .on('zoom', (e: D3ZoomEvent<SVGGElement, unknown>) => {
+        const isZoomClick = this.isZoomClick
+        this.draw = true
+        this.isZoomClick = false
+
+        return this.container
+          .transition()
+          .duration(isZoomClick ? 400 : 20)
+          .call(sel => (isZoomClick ? sel.ease(easeCubic) : sel))
+          .attr('transform', String(e.transform))
+      })
```

Comment:
Can this be extracted to an event handler file? It's not directly related to visualisation itself.

Reply:
I tried but it was a bit tricky since it's referencing some class properties (draw and isZoomClick) that are changed by some of the other methods. I could in theory pass `this` to it, but it didn't feel like a clean solution, so I left as is for now.